### PR TITLE
MAILBOX-314 elasticsearch: do not split email when tokenizing WIP

### PR DESCRIPTION
**Pull request by strokyl**

### CODE REVIEW

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 1]
```diff
@@ -1051,7 +1080,7 @@ public void searchWithTextShouldReturnMailsWhenCcMatches() throws Exception {
     @Test
     public void searchWithTextShouldReturnMailsWhenBccMatches() throws Exception {
         Assume.assumeTrue(storeMailboxManager.getSupportedSearchCapabilities().contains(MailboxManager.SearchCapabilities.Text));
-        SearchQuery searchQuery = new SearchQuery(SearchQuery.textContains("any.com"));
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.textContains("any@any.com"));
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: so it's no more possible to search for any to find any@example.com? And so search for a domain to have all emails related to a customer?

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: We may achieve this with search analyzer and index analyzer. 

If we index using standard analyzer on a first field and email analyzer on another one, then execute or search query on both with only email analyzer, I guess it works.

**Comment by strokyl** [comment]
push -f because rebase on master

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    protected TextExtractor getTextExtractor() {
+        return new DefaultTextExtractor();
+    }
+
+    @Override
+    @Test
+    public void searchWithPDFAttachmentShouldReturnMailsWhenAttachmentContentMatches() throws Exception {
+        Assume.assumeTrue(false);
```
Comment (strokyl) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java]: I guess there is something better we can do here "@Ignore" did not work

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 3]
```diff
@@ -751,6 +780,22 @@ public void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecified() thro
...
     @Test
+    public void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithOnlyUserPartOfEmail() throws Exception {
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.From, "murari"));
+
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m8.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithDomainPartOfEmail() throws Exception {
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: Don't work with Lucene

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 4]
```diff
@@ -759,8 +804,53 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
...
     @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecifiedWithOnlyEmailUserPart() throws Exception {
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.To, "root"));
+
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m1.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecifiedWithOnlyDomainPartSpecified() throws Exception {
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: Don't work with Lucene

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 5]
```diff
@@ -759,8 +804,53 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
...
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenCcIsSpecifiedWithOnlyUserPartOfTheEmail() throws Exception {
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.Cc, "monkey"));
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m5.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenCcIsSpecifiedWithOnlyDomainPartOfTheEmail() throws Exception {
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: Don't work with Lucene

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 6]
```diff
@@ -759,8 +804,53 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
...
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenBccIsSpecifiedWithOnlyUserPartOfTheEmail() throws Exception {
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.Bcc, "monkey"));
+
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m5.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenBccIsSpecifiedWithOnlyDomainPartOfTheEmail() throws Exception {
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: Don't work with Lucene

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java (hunk 2 above)
Comment (rouazana) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java]: generally @Ignore + @Override + empty test body works and is pretty clear about the intent

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java [hunk 7]
```diff
@@ -35,6 +35,9 @@
     private static final int DEFAULT_NB_SHARDS = 1;
     private static final int DEFAULT_NB_REPLICA = 0;
     public static final String CASE_INSENSITIVE = "case_insensitive";
+    public static final String KEEP_MAIL_AND_URL = "keep_mail_and_url";
+    public static final String SNOWBALL_KEEP_MAIL_AND_URL = "snowball_keep_mail_and_token";
```
Comment (rouazana) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: s/AND_URL/AND_TOKEN/ ?
It's very hard from these constant name to know what there are used for.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java [hunk 8]
```diff
@@ -19,10 +19,10 @@
 
 package org.apache.james.mailbox.elasticsearch;
 
-    import org.apache.james.mailbox.extractor.TextExtractor;
-    import org.apache.james.mailbox.store.extractor.DefaultTextExtractor;
-    import org.junit.Assume;
-    import org.junit.Test;
```
Comment (rouazana) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java]: should be a fixup of previous commit

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 9]
```diff
@@ -1033,7 +1123,15 @@ public void searchWithTextShouldReturnMailsWhenFromMatches() throws Exception {
...
     public void searchWithTextShouldReturnMailsWhenToMatches() throws Exception {
         Assume.assumeTrue(storeMailboxManager.getSupportedSearchCapabilities().contains(MailboxManager.SearchCapabilities.Text));
-        SearchQuery searchQuery = new SearchQuery(SearchQuery.textContains("listes.minet.net"));
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.textContains("root@listes.minet.net"));
+
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsExactly(m1.getUid());
+    }
+    @Test
+    public void searchWithTextShouldReturnMailsWhenToIsNotAExactMatches() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: s/AExact/AnExact/

File: mailbox/store/src/test/resources/eml/mail5.eml [hunk 10]
```diff
@@ -0,0 +1,52 @@
+Return-Path: <mailet-api-return-812-benwa=minet.net@james.apache.org>
+Received: from mx2.minet.net (mx2.minet.net [192.168.102.26])
+	 by imap (Cyrus v2.4.16-Debian-2.4.16-4+deb7u1) with LMTPA;
+	 Fri, 15 May 2015 08:36:03 +0200
+X-Sieve: CMU Sieve 2.4
+Received: from localhost (spam.minet.net [192.168.102.97])
+	by mx2.minet.net (Postfix) with ESMTP id 5AC87A19A4F
+	for <benwa@minet.net>; Fri, 15 May 2015 08:36:03 +0200 (CEST)
```
Comment (rouazana) [inline mailbox/store/src/test/resources/eml/mail5.eml]: anonymization?

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/AbstractElasticIntegrationTest.java [hunk 11]
```diff
@@ -76,7 +72,7 @@
 import com.google.common.base.Charsets;
 import com.google.common.base.Strings;
 
-public class ElasticSearchIntegrationTest extends AbstractMessageSearchIndexTest {
+public abstract class AbstractElasticIntegrationTest extends AbstractMessageSearchIndexTest {
```
Comment (aduprat) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/AbstractElasticIntegrationTest.java]: multiple `abstract` level, you win ! ;)

File: mailbox/store/src/test/resources/eml/mail5.eml (hunk 10 above)
Comment (strokyl) [reply mailbox/store/src/test/resources/eml/mail5.eml]: Benwa is already everywhere (almost all .eml). It's a nickname not a real name. Benoit tell me it's does care.
If you think it is important create a ticket for that I won't do that in that PR

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java (hunk 7 above)
Comment (strokyl) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: I know but could not find other name, I am open to suggestion!

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 1 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: DONE

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java (hunk 2 above)
Comment (strokyl) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java]: I found something better !

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 3 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I add a search capability that lucene does not implement to solve that issue

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 4 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I add a search capability that lucene does not implement to solve that issue

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 5 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I add a search capability that lucene does not implement to solve that issue

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 6 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I add a search capability that lucene does not implement to solve that issue

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java (hunk 7 above)
Comment (strokyl) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: I am open to suggestion

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java (hunk 8 above)
Comment (strokyl) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTestWithDefaultTextExtractor.java]: That should be fixed when doing last rebase --autosquash before merging

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 9 above)
Comment (strokyl) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: done

**Comment by strokyl** [comment]
test this please

**Comment by strokyl** [comment]
Closing this PR too rerun test...

**Comment by strokyl** [comment]
New PR here: https://github.com/linagora/james-project/pull/1078

**Pull request closed without merging**